File: creational_patterns/java/simple_factory/SimpleFactory.java
package creational_patterns.java.simple_factory;
import java.util.function.*;
import java.util.*;;

// Log types as enums
enum LogType{
    ERROR, DEBUG, INFO
}

// ILogger interface for abstaction
interface ILogger{
    void log(String message);
}

// Concrete logger classes
class ErrorLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[ERROR]: " + message);
    }
}

class DebugLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[DEBUG]: " + message);
    }
}

class InfoLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[INFO]: " + message);
    }
}

// Logger factory - creates a logger object based on log type.
class LoggerFactory{

    private final static Map<LogType, Supplier<ILogger>> store = new HashMap<>();

    static {
        register(LogType.DEBUG, DebugLogger::new);
        register(LogType.ERROR, ErrorLogger::new);
        register(LogType.INFO, InfoLogger::new);
    }

    private static void register(LogType logType, Supplier<ILogger> supplier){
        store.put(logType, supplier);   
    }

    public ILogger createLogger(LogType logType){
        Supplier<ILogger> supplier = store.get(logType);
        if( supplier != null ){
            return supplier.get();
        }
        return null;
    }

    // Voilates OCP
    // public ILogger createLogger(LogType logType){
    //     switch (logType){
    //         case INFO:
    //             return new InfoLogger();
    //         case DEBUG:
    //             return new DebugLogger();
    //         case ERROR:
    //             return new ErrorLogger();
    //         default:
    //             return new InfoLogger();
    //     }
    // }
}

// Client
class SimpleFactory{
    public static void main(String args[]){
        LoggerFactory loggerFactory = new LoggerFactory();

        // Specific Loggers
        ILogger logger;
        logger = loggerFactory.createLogger(LogType.DEBUG);
        logger.log("Hola..!");

        logger = loggerFactory.createLogger(LogType.ERROR);
        logger.log("Hola..!");
    }
}

File: creational_patterns/c++/simple_factory/SimpleFactory.cpp
#include<iostream>
#include<string>
#include<memory>
#include<functional>
using namespace std;

class ILogger{
public:
    virtual void log(string message) = 0;
    virtual ~ILogger() {}
};

class ErrorLogger: public ILogger {
public:
    void log(string message) override {
        cout << "[ERROR]: " << message << endl;
    }
};

class DebugLogger: public ILogger {
public:
    void log(string message) override {
        cout << "[DEBUG]: " << message << endl;
    }
};

class InfoLogger: public ILogger {
public:
    void log(string message) override {
        cout << "[INFO]: " << message << endl;
    }
};

enum class LoggerType{
    ERROR, INFO, DEBUG
};

class LoggerFactory {
private:
    using Creator = function<unique_ptr<ILogger>()>;
    unordered_map<LoggerType, Creator> mp;

public:

    LoggerFactory(){
        registerLogger(LoggerType::ERROR, [](){return make_unique<ErrorLogger>(); });
        registerLogger(LoggerType::DEBUG, [](){return make_unique<DebugLogger>(); });
        registerLogger(LoggerType::INFO, [](){return make_unique<InfoLogger>(); });
    }

    void registerLogger(LoggerType loggerType, Creator creator){
        mp[loggerType] = creator;
    }

    unique_ptr<ILogger> createLogger(LoggerType type){
        auto it = mp.find(type);
        if( it != mp.end() ){
            return it->second();
        }
        return nullptr;
    }
};

int main(){
    LoggerFactory loggerFactory;
    unique_ptr<ILogger> logger = loggerFactory.createLogger(LoggerType::ERROR);
    if ( logger ){
        logger->log("Hello Vineel");
    }else{
        cout << "Failed to create logger" << endl;
    }
}